src/components/EmailNotifications/EmailNotificationsProvider.js

```diff
@@ -0,0 +1,142 @@
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react'
+import PropTypes from 'prop-types'
+import { useWallet } from '../../providers/Wallet'
+import {
+  getJurorEmail,
+  getSubscriptionDetails,
+  subscribeExistingEmail,
+  subscribeToNotifications,
+} from '../../services/servicesRequests'
+
+const EmailNotificationsContext = React.createContext()
+
+function EmailNotificationsProvider({ children }) {
+  const { account } = useWallet()
+
+  const asyncCancelled = useRef(false)
+  const [email, setEmail] = useState(null)
+  const [needsSignature, setNeedsSignature] = useState(false)
+  const [subscriptionDetails, setSubscriptionDetails] = useState({
+    error: null,
+    fetching: false,
+  })
+
+  const handleOnSubscribe = useCallback(
+    async email => {
+      const response = await subscribeToNotifications(account, email)
+
+      if (response.error && !response.needsSignature) {
+        return response.error
+      }
+
+      if (!asyncCancelled.current) {
+        setEmail(email)
+        if (needsSignature !== Boolean(response.needsSignature)) {
+          setNeedsSignature(!needsSignature)
+        }
+      }
+    },
+    [account, needsSignature]
+  )
+
+  const handleOnSubscribeExistingEmail = useCallback(async () => {
+    const response = await subscribeExistingEmail(account)
+
+    if (response.error && !response.needsSignature) {
+      return response.error
+    }
+
+    if (!asyncCancelled.current) {
+      if (email !== response.email) {
+        setEmail(response.email)
+        if (needsSignature !== Boolean(response.needsSignature)) {
+          setNeedsSignature(!needsSignature)
+        }
+      }
+    }
+  }, [account, email, needsSignature])
+
+  // Cancel any async requests if this provider is unmounted
+  useEffect(() => {
+    return () => {
+      asyncCancelled.current = true
+    }
+  }, [])
+
+  // When account connects, fetch their subscription details
+  useEffect(() => {
+    if (!account) {
+      return
+    }
+
+    const fetchSubscriptionDetails = async () => {
+      const response = await getSubscriptionDetails(account)
+
+      if (!asyncCancelled.current) {
+        setSubscriptionDetails({
+          ...response,
+          fetching: false,
+        })
+      }
+    }
+
+    setSubscriptionDetails(subscriptionDetails => ({
+      ...subscriptionDetails,
+      error: null,
+      fetching: true,
+    }))
+    fetchSubscriptionDetails()
+  }, [account])
+
+  // Once we know the account has an associated email, fetch its email
+  useEffect(() => {
+    if (!subscriptionDetails.emailExists) {
+      return
+    }
+
+    const getEmail = async () => {
+      const { needsSignature, email } = await getJurorEmail(account)
```

Comment:
Giving a second thought to this, i think that here we are going to be fetching the juror email every time an account is connected, and when the cookie has expired we are going to ask the juror to sign a message just for connecting his account and not sure if that is what we want.
The way i see this is that we need to do that only when we are subscribing to notifications through the modal or entering the global preferences not every time the juror connect his account to interact with the dashboard.
What do you think?

Reply:
Good point. Yes, it sounds like we should do this in a callback instead.